### ISSUE TITLE
[Bug]: sorting does not works in ASR train

### ISSUE THREAD

**Issue by starcell** [speechbrain/speechbrain#1722]
### Describe the bug

When training with mult-GPU.
Setting sorting: random and sorting: ascending in hparams/*.yaml file show the same activity. 
In case sorting setted ascending, batch dataset processed randomly in training. 

### Expected behaviour

When I set sorting=ascending, shorter data trained earlier.

### To Reproduce

In yaml file, set as below
sorting: ascending

then ASR train!
It takes the same time when set sorting: random

### Versions

v0.5.13

With v0.5.11, no problem appeared.
I can train with short data at first in v0.5.11.

### Additional context

I tested and come to know there are some probelms in line 1000 of core.py.

1000        with tqdm(
                    train_set,
                    initial=self.step,
                    dynamic_ncols=True,
                    disable=not enable,
                ) as t:
                    for batch in t:

the train_set is sorted befor tqdm, but batch data is not sorted after tqdm().

**Comment by starcell** [comment]
Solution!
I found where is wrong.
The value of 'shuffle' on the line 760 of core.py must be 'Fasle', not 'True'.
dataset, rank=self.rank, shuffle=False, drop_last=drop_last

I wish source code checked and corrected.

**Comment by AsuMagic** [comment]
For the record, this change was introduced by #1518.

**Comment by AsuMagic** [comment]
My understanding of the current code is that `shuffle=True` is passed to the [DistributedSampler](https://pytorch.org/docs/stable/data.html#torch.utils.data.distributed.DistributedSampler) so that each DDP instance receives random subsets of the dataset at each epoch.  

Assuming you are correct, what I'm assuming is happening and may be worth digging into:  
To my understanding, sorting is happening in `dataio_prepare`, before the dataset sampling.  
But the sampler's `shuffle=True` does not only shuffle what subsets DDP instances receives, it also shuffles the order of indices within that sample, which breaks the sorting order.

This is the relevant snippet of PyTorch code:

```python
    def __iter__(self) -> Iterator[T_co]:
        if self.shuffle:
            # deterministically shuffle based on epoch and seed
            g = torch.Generator()
            g.manual_seed(self.seed + self.epoch)
            indices = torch.randperm(len(self.dataset), generator=g).tolist()  # type: ignore[arg-type]
        else:
            indices = list(range(len(self.dataset)))  # type: ignore[arg-type]

        if not self.drop_last:
            # add extra samples to make it evenly divisible
            padding_size = self.total_size - len(indices)
            if padding_size <= len(indices):
                indices += indices[:padding_size]
            else:
                indices += (indices * math.ceil(padding_size / len(indices)))[:padding_size]
        else:
            # remove tail of data to make it evenly divisible.
            indices = indices[:self.total_size]
        assert len(indices) == self.total_size

        # subsample
        indices = indices[self.rank:self.total_size:self.num_replicas]
        assert len(indices) == self.num_samples

        return iter(indices)
```

I don't know what a proper way of fixing this would be. It might be a little silly but could sorting the indices obtained from the sampler work...?

**Comment by TParcollet** [comment]
Hum, interesting. @RuABraun @Gastron, any idea?

**Comment by TParcollet** [comment]
Maybe it is fine to have a fixed order between GPUs if sorting != random. Should we simply add a check and set it back to False if sorting != random?

**Comment by RuABraun** [comment]
My bad for assuming one would always want random shuffling. I agree with the understanding of @AsuMagic . 

I think it would make sense to have any sorting that is done to happen inside the sampler, after the shuffle, and after line `indices = indices[self.rank:self.total_size:self.num_replicas]`. Instead of in `dataio_prepare`. This way the sorting logic would be contained in one location, instead of having code in different locations be simultaneously responsible (which I think makes it harder to comprehend).

Alternatively, adding an `if sorting != random` then `shuffle=False` would be a quick fix that would work. However I do think it's better for performance to have different splits for each epoch (so this solution would not be ideal in my opinion).

**Comment by TParcollet** [comment]
@anautsch @Adel-Moumen this is absolutely critical as it impacts all recipe using ascending / descending. We should solve this asap please.

**Comment by anautsch** [comment]
Sounds like a mix of topics colliding:
* core <> dataio behaviour vs pytorch
* ddp vs single core => ddp won't be testable as such
* tqdm & pytorch dependency & changes we can track only through testing
* implications to: creating/using shards & dynamic batching

Aside the w2v2 pretraining vs all other recipes.

@AsuMagic & @RuABraun bring up some valid points. Yet, should we really create our own sorting implementation and ignore simply other options for sorting?

DDP, shards & dynamic batching all need to deal with the same issues on availing randomness while pre-partitioning the dataset (which batching is itself also does). It might sound stupid, but is there some "standard" literature on this? The core concept is to sustain properties of a Randomized Controlled Trial (treating all unknowns as aleatory uncertainty only), and nothing more, or?

**Comment by anautsch** [comment]
How about undoing 1518 and re-randomising data when wav2vec2 needs it?
(Not a w2v2 expert...)

**Comment by AsuMagic** [comment]
I imagine the issue raised by @RuABraun is not necessarily specific to wav2vec2, and _could_ impact other existing recipes in SB?

I don't know if moving the dataset sorting logic after the sampling will work, but it might be an option. However, the existence of samplers like `DynamicBatchSampler` and other user-defined samplers must be considered.

It might be reinventing the wheel, but maybe reimplementing `DistributedSampler` in SB is the way to go? It could allow more flexibility on the long-term anyway.  
I still think sorting the indices is the most straightforward option if we don't want to move all the sorting logic away from `dataio_prepare`. Reimplementing the sampler would make this rather straightforward and efficient (as we could do it at tensor level). A hyperparameter could specify whether we want the DDP slices to be random per-epoch, per-seed, or not random at all.

One possibly elegant approach would be to rewrite `DistributedSamplerWrapper` to include the distributed sampling logic, instead of inheriting from PyTorch's `DistributedSampler`.  
The equivalent to the current DDP default would be `DistributedSamplerWrapper(ReproducibleRandomSampler())`.  
FWIW, no recipe directly makes use of PyTorch's `DistributedSampler`, all references are internal SB defaults.

**Comment by anautsch** [comment]
I lack the time to refactor sampling as well before xmas (recipe testing & HF transformer interface). I agree with you: this needs to be studied, a plan outlined, implementation happening with testing, and refactored accompanied with a tutorial.

@TParcollet suggested (per pm) to roll out a solution now, so that all recipes expecting sorting are delivering sorting.
I get that, too. Time is critical for researchers, partially whose theses depend on this being correct as much as it can/should be expected to be correct.

Started reading into randomized controlled trials at hospitals and how they'd handle data splits. @AsuMagic just as you mention - we need testing here; serious testing. It's just so critical. Yet, what to test for?

A two stage approach could be good: desinfect the wound & bandage it - then take time for the healing process.

The refactoring necessary to solve this to the extent that needs - throughout all of SB - it's at least a month start-to-finish.

**Comment by AsuMagic** [comment]
> A two stage approach could be good: desinfect the wound & bandage it - then take time for the healing process.

Agreed, the intended benefit is minor compared to sorting being broken right now anyway.  
I assume the goal is to push a SB hotfix with this sorted out ASAP?

> Started reading into randomized controlled trials at hospitals and how they'd handle data splits. @AsuMagic just as you mention - we need testing here; serious testing. It's just so critical. Yet, what to test for?

Reasoning from the breakage induced by the PR, I think using some test data to validate whether the batches gone through as much of the process as possible (sorting, sampling, etc.) follow the desired parameters in different test settings would be a good start.  
You mentioned recipe testing: Maybe some ordering/sampling/etc. sanity tests would fit here?

> Having to redo the sampler makes me rethink my earlier comment: If it really is just about supporting ascending and random, then maybe the fix of just having an if!=random -> shuffle=False, (plus possibly a onetime warning that the user is responsible for changing ordering between epochs) is enough. Rewriting the sampler seems like overkill for just having the capability to do ascending/descending/random.

Agreed on the "If it really is just about supporting ascending and random [...]" part.  
The question is indeed whether random epoch splits are really all that important in a sorted case. IMO, if it helps "a bit", I think this slight tradeoff in code complexity is worth to have.

To clarify, my point of view is that the current flow involving sorting the dataset at data preparation time is fine. It also makes sense for the dataset sampling code to be the code in charge of sampling dataset slices that are random per-epoch.  
Hence, fixing said dataset sampling code where it's "broken" makes sense to me even if forking the distributed sampler code from PyTorch isn't an ideal solution: PyTorch's code makes the assumption that we don't care about the ordering when requesting random slices when we in fact do.

TBH, I don't know if it is bikeshedding or if I'm repeating myself accidentally too much but dataset sorting seems critical enough that a robust solution is important :)

**Comment by anautsch** [comment]
@AsuMagic it's ok to repeat oneself here and there - sometimes it aids clarity and understanding one another's point. We're discussing here; not writing scripture.

Thank you for bringing up a series of good points. They help to structure the particular next steps, and outline what one can do afterwards—dynamic batching has lots of potential, still.

This is a curiosity question: 
How do we care about the ordering when requesting random slices in SB while PyTorch might not do so?

**Comment by AsuMagic** [comment]
> This is a curiosity question: How do we care about the ordering when requesting random slices in SB while PyTorch might not do so?

Assuming you're talking about the same thing as me, i.e. the `DistributedSampler` implementation, what it does is to:
1. Generate a tensor of indices from `0` to `len(dataset)-1`
2. Shuffle the tensor of indices
3. Converts the tensor of indices to a list
4. Slice into the list depending on the DDP instance we are running on

So with starting indices of `[0, 1, 2, 3, 4, 5, 6, 7]`, you might end up with `[4, 1, 2, 3, 5, 6, 7, 0]`. With two DDP instances, one DDP instance would end up with `[4, 2, 5, 7]` and the other with `[1, 3, 6, 0]` (if I got the slicing logic right).  
As a result, your slice has shuffled indices. If you were to build batches of 2 from these (which is what seemingly causes the issue in SB), on the second instance, your first batch would use indices `[1, 3]`, which is not the actual order of data in the base dataset.

The solution I was talking about is to sort these sliced indices, so you would obtain `[2, 4, 5, 7]` for the first instance and `[0, 1, 3, 6]` for the second. This would restore the relative element order, while still picking slices at random (deterministically for a given seed+epoch, though), which _might_, at least for some recipes, improve accuracy.

This one specific approach can be achieved in two ways:
- Either in SB after sampling from PyTorch's `DistributedSampler`
- Either by reimplementing `DistributedSampler` and integrating this sorting logic within it (or rewriting the sampling logic so that it doesn't shuffle elements within slices in the first place, but I don't think it'd actually be better).

The former approach is a workaround, but it'd probably work.  
The latter implies adding more code to SB, but it might be somewhat more efficient.

**Comment by anautsch** [comment]
The above mentioned 1730 PR has been merged. Closing here.
* fixed the DDP & sorting bug
* provided an integration test
* made SB ready for cpu-only DDP (necessary for regular testing with github workflows)

Please feel free to re-open, when needed.
